src/core/utils/jam_detector.hpp

```diff
@@ -0,0 +1,206 @@
+     * @retval kThreadErrorAlready      Jam detection has been started before.
+     *
+     */
+    ThreadError Start(Handler aHandler, void *aContext);
+
+    /**
+     * Stop the jamming detection.
+     *
+     * @retval kThreadErrorNone         Successfully stopped the jamming detection.
+     * @retval kThreadErrorAlready      Jam detection is already stopped.
+     *
+     */
+    ThreadError Stop(void);
+
+    /**
+     * Get the Jam Detection Status
+     *
+     * @returns The Jam Detection status (true if enabled, false otherwise).
+     */
+    bool IsEnabled(void) const { return mEnabled; }
+
+    /**
+     * Get the current jam state.
+     *
+     * @returns The jamming state (`true` if jam is detected, `false` otherwise).
+     */
+    bool GetState(void) const { return mJamState; }
+
+    /**
+     * Set the Jam Detection RSSI Threshold (in dBm).
+     *
+     * @param[in]  aRssiThreshold  The RSSI threshold.
+     *
+     * @retval kThreadErrorNone    Successfully set the threshold.
+     *
+     */
+    ThreadError SetRssiThreshold(int8_t aThreshold);
+
+    /**
+     * Get the Jam Detection RSSI Threshold (in dBm).
+     *
+     * @returns The Jam Detection RSSI Threshold.
+     */
+    int8_t GetRssiThreshold(void) const { return mRssiThreshold; }
+
+    /**
+     * Set the Jam Detection Detection Window (in seconds).
+     *
+     * @param[in]  aWindow              The Jam Detection window (valid range is 1 to 63)
+     *
+     * @retval kThreadErrorNone         Successfully set the window.
+     * @retval kThreadErrorInvalidArgs  The given input parameter not within valid range (1-63)
+     *
+     */
+    ThreadError SetWindow(uint8_t aWindow);
+
+    /**
+     * Get the Jam Detection Detection Window (in seconds).
+     *
+     * @returns The Jam Detection Window.
+     */
+    uint8_t GetWindow(void) const { return mWindow; }
+
+    /**
+     * Set the Jam Detection Busy Period (in seconds).
+     *
+     * The number of aggregate seconds within the detection window where the RSSI must be above
+     * threshold to trigger detection.
+     *
+     * @param[in]  aBusyPeriod          The Jam Detection busy period (should be non-zero and
+                                        less than or equal to Jam Detection Window)
+     *
+     * @retval kThreadErrorNone         Successfully set the window.
+     * @retval kThreadErrorInvalidArgs  The given input is not within the valid range.
+     *
+     */
+    ThreadError SetBusyPeriod(uint8_t aBusyPeriod);
+
+    /**
+     * Get the Jam Detection Busy Period (in seconds)
+     *
+     * @returns The Jam Detection Busy Period
+     */
+    uint8_t GetBusyPeriod(void) const { return mBusyPeriod; }
+
+private:
+    static void HandleTimer(void *aContext);
+    void HandleTimer(void);
+    void UpdateHistory(bool aThresholdExceeded);
+    void UpdateJamState(void);
+
+private:
+    enum
+    {
+        kMaxWindow              = 63,
```

Comment:
This is what we chose as the API limit. We allow the window range to be `1-63` seconds.  This is a reasonable range and if there is a need to for larger range this can be changed/updated.
The reason for this limit is to be able to use  `uint64_t` a history bitmap